File: grok/clients/python/grok_client.py
"""
xAI Grok Client for Python — DevSkyy
Grok uses OpenAI-compatible API at https://api.x.ai/v1
Unique features: Live web search, real-time data, X/Twitter access
"""

import os
import json
import time
import base64
from pathlib import Path
from typing import Optional, List, Dict, Any, Generator
import openai
from dotenv import load_dotenv

load_dotenv(Path(__file__).parent.parent.parent / '.env')


class GrokClient:
    """xAI Grok client — OpenAI SDK pointed at api.x.ai"""

    def __init__(self, api_key: Optional[str] = None, **options):
        self.api_key = api_key or os.getenv('XAI_API_KEY') or os.getenv('GROK_API_KEY')
        if not self.api_key:
            raise ValueError('xAI API key required. Set XAI_API_KEY in .env')

        settings_path = Path(__file__).parent.parent / 'config' / 'settings.json'
        with open(settings_path) as f:
            self.settings = json.load(f)

        models_path = Path(__file__).parent.parent / 'config' / 'models.json'
        with open(models_path) as f:
            self.model_configs = json.load(f)

        self.client = openai.OpenAI(
            api_key=self.api_key,
            base_url=options.get('base_url', self.settings['baseURL'])
        )

        self.default_model = options.get('model', self.settings['defaultModel'])
        self.generation_config = {
            **self.settings['generationConfig'],
            **options.get('generation_config', {})
        }
        self.last_request_time = 0.0
        self.min_request_interval = 60.0 / self.settings['rateLimit']['requestsPerMinute']

    def _rate_limit(self):
        now = time.time()
        elapsed = now - self.last_request_time
        if elapsed < self.min_request_interval:
            time.sleep(self.min_request_interval - elapsed)
        self.last_request_time = time.time()

    def generate_content(self, prompt: str, model: Optional[str] = None,
                         system_prompt: Optional[str] = None, **config) -> Dict[str, Any]:
        self._rate_limit()
        messages = []
        if system_prompt:
            messages.append({'role': 'system', 'content': system_prompt})
        messages.append({'role': 'user', 'content': prompt})

        try:
            response = self.client.chat.completions.create(
                model=model or self.default_model,
                messages=messages,
                **{**self.generation_config, **config}
            )
            return {
                'text': response.choices[0].message.content,
                'finish_reason': response.choices[0].finish_reason,
                'usage': {
                    'prompt_tokens': response.usage.prompt_tokens,
                    'completion_tokens': response.usage.completion_tokens,
                    'total_tokens': response.usage.total_tokens
                }
            }
        except Exception as e:
            return self._handle_error(e)

    def generate_content_stream(self, prompt: str, model: Optional[str] = None,
                                system_prompt: Optional[str] = None, **config) -> Generator:
        self._rate_limit()
        messages = []
        if system_prompt:
            messages.append({'role': 'system', 'content': system_prompt})
        messages.append({'role': 'user', 'content': prompt})

        try:
            with self.client.chat.completions.stream(
                model=model or self.default_model,
                messages=messages,
                **{**self.generation_config, **config}
            ) as stream:
                for chunk in stream:
                    delta = chunk.choices[0].delta.content if chunk.choices else ''
                    if delta:
                        yield {'text': delta, 'done': False}
            yield {'text': '', 'done': True}
        except Exception as e:
            raise self._handle_error(e)

    def live_search(self, query: str, model: str = 'grok-3',
                    sources: Optional[List] = None,
                    system_prompt: str = 'You have access to real-time web and X data.') -> Dict[str, Any]:
        """Grok live search — real-time web + X/Twitter grounding"""
        self._rate_limit()
        if sources is None:
            sources = [{'type': 'web'}, {'type': 'x'}]

        try:
            response = self.client.chat.completions.create(
                model=model,
                messages=[
                    {'role': 'system', 'content': system_prompt},
                    {'role': 'user', 'content': query}
                ],
                search_parameters={'mode': 'on', 'sources': sources},
                max_tokens=self.generation_config.get('max_tokens', 2048)
            )
            msg = response.choices[0].message
            return {
                'text': msg.content,
                'citations': getattr(msg, 'citations', []),
                'usage': response.usage
            }
        except Exception as e:
            return self._handle_error(e)

    def analyze_image(self, image_path: Optional[str] = None, image_url: Optional[str] = None,
                      prompt: str = 'Describe this image', model: str = 'grok-2-vision-1212',
                      detail: str = 'auto') -> Dict[str, Any]:
        self._rate_limit()
        if image_path:
            with open(image_path, 'rb') as f:
                data = base64.b64encode(f.read()).decode()
            ext = Path(image_path).suffix.lstrip('.').replace('jpg', 'jpeg')
            image_content = {'type': 'image_url',
                             'image_url': {'url': f'data:image/{ext};base64,{data}', 'detail': detail}}
        elif image_url:
            image_content = {'type': 'image_url', 'image_url': {'url': image_url, 'detail': detail}}
        else:
            raise ValueError('Either image_path or image_url required')

        try:
            response = self.client.chat.completions.create(
                model=model,
                messages=[{'role': 'user', 'content': [
                    {'type': 'text', 'text': prompt}, image_content
                ]}],
                max_tokens=self.generation_config.get('max_tokens', 1024)
            )
            return {'text': response.choices[0].message.content, 'usage': response.usage}
        except Exception as e:
            return self._handle_error(e)

    def _handle_error(self, error: Exception) -> Dict[str, Any]:
        response = {'error': True, 'message': str(error), 'type': type(error).__name__}
        msg = str(error).lower()
        if '401' in msg or 'authentication' in msg:
            response.update(type='AuthenticationError', message='Invalid API key. Check XAI_API_KEY')
        elif '429' in msg or 'rate limit' in msg:
            response.update(type='RateLimitError', message='Rate limit exceeded')
        if self.settings['logging']['logRequests']:
            print(f'Grok API Error: {response}')
        raise Exception(response)

    def get_available_models(self) -> List[Dict]:
        return self.model_configs['models']

    def get_recommended_model(self, task_type: str) -> str:
        return self.model_configs['modelSelection'].get(task_type, self.default_model)

File: grok/clients/node/grok-client.js
/**
 * xAI Grok Client for Node.js — DevSkyy
 * Grok uses OpenAI-compatible API at https://api.x.ai/v1
 * Unique features: Live web search, real-time data, X/Twitter access
 */

const OpenAI = require('openai');
const fs = require('fs');
const path = require('path');
require('dotenv').config({ path: path.join(__dirname, '../../../.env') });

class GrokClient {
  constructor(options = {}) {
    this.apiKey = options.apiKey || process.env.XAI_API_KEY || process.env.GROK_API_KEY;

    if (!this.apiKey) {
      throw new Error('xAI API key required. Set XAI_API_KEY in .env');
    }

    const settingsPath = path.join(__dirname, '../../config/settings.json');
    this.settings = JSON.parse(fs.readFileSync(settingsPath, 'utf8'));

    const modelsPath = path.join(__dirname, '../../config/models.json');
    this.modelConfigs = JSON.parse(fs.readFileSync(modelsPath, 'utf8'));

    // Grok uses OpenAI SDK with xAI base URL
    this.client = new OpenAI({
      apiKey: this.apiKey,
      baseURL: options.baseURL || this.settings.baseURL
    });

    this.defaultModel = options.model || this.settings.defaultModel;
    this.generationConfig = { ...this.settings.generationConfig, ...options.generationConfig };

    this.lastRequestTime = 0;
    this.minRequestInterval = 60000 / this.settings.rateLimit.requestsPerMinute;
  }

  async rateLimit() {
    const now = Date.now();
    const elapsed = now - this.lastRequestTime;
    if (elapsed < this.minRequestInterval) {
      await new Promise(r => setTimeout(r, this.minRequestInterval - elapsed));
    }
    this.lastRequestTime = Date.now();
  }

  /** Generate content */
  async generateContent(options) {
    await this.rateLimit();
    const { prompt, model, systemPrompt, ...config } = options;

    const messages = [];
    if (systemPrompt) messages.push({ role: 'system', content: systemPrompt });
    messages.push({ role: 'user', content: prompt });

    try {
      const response = await this.client.chat.completions.create({
        model: model || this.defaultModel,
        messages,
        ...this.generationConfig,
        ...config
      });

      return {
        text: response.choices[0].message.content,
        finishReason: response.choices[0].finish_reason,
        usage: response.usage
      };
    } catch (error) {
      return this.handleError(error);
    }
  }

  /** Streaming */
  async *generateContentStream(options) {
    await this.rateLimit();
    const { prompt, model, systemPrompt, ...config } = options;

    const messages = [];
    if (systemPrompt) messages.push({ role: 'system', content: systemPrompt });
    messages.push({ role: 'user', content: prompt });

    try {
      const stream = await this.client.chat.completions.create({
        model: model || this.defaultModel,
        messages,
        stream: true,
        ...this.generationConfig,
        ...config
      });

      for await (const chunk of stream) {
        const delta = chunk.choices[0]?.delta?.content || '';
        if (delta) yield { text: delta, done: false };
      }
      yield { text: '', done: true };
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /** Multi-turn chat */
  startChat(options = {}) {
    const { model, systemPrompt = '', history = [] } = options;
    const client = this.client;
    const defaultModel = model || this.defaultModel;
    const genConfig = this.generationConfig;

    return {
      model: defaultModel,
      history: systemPrompt
        ? [{ role: 'system', content: systemPrompt }, ...history]
        : [...history],

      async sendMessage(message) {
        this.history.push({ role: 'user', content: message });
        const response = await client.chat.completions.create({
          model: this.model, messages: this.history, ...genConfig
        });
        const reply = response.choices[0].message.content;
        this.history.push({ role: 'assistant', content: reply });
        return { response: { text: () => reply }, usage: response.usage };
      }
    };
  }

  /**
   * Live search — Grok's signature feature
   * Accesses real-time web + X/Twitter data
   */
  async liveSearch(options) {
    await this.rateLimit();
    const {
      query,
      model = 'grok-3',
      sources = [{ type: 'web' }, { type: 'x' }],
      systemPrompt = 'You are a helpful assistant with access to real-time web and X (Twitter) data.'
    } = options;

    try {
      const response = await this.client.chat.completions.create({
        model,
        messages: [
          { role: 'system', content: systemPrompt },
          { role: 'user', content: query }
        ],
        // xAI search grounding extension
        search_parameters: { mode: 'on', sources },
        max_tokens: this.generationConfig.max_tokens
      });

      const msg = response.choices[0].message;
      return {
        text: msg.content,
        citations: msg.citations || [],
        usage: response.usage
      };
    } catch (error) {
      return this.handleError(error);
    }
  }

  /** Vision — image analysis with Grok 2 Vision */
  async analyzeImage(options) {
    await this.rateLimit();
    const { imagePath, imageUrl, prompt, model = 'grok-2-vision-1212', detail = 'auto' } = options;

    let imageContent;
    if (imagePath) {
      const data = fs.readFileSync(imagePath);
      const ext = path.extname(imagePath).slice(1).replace('jpg', 'jpeg');
      imageContent = {
        type: 'image_url',
        image_url: { url: `data:image/${ext};base64,${data.toString('base64')}`, detail }
      };
    } else if (imageUrl) {
      imageContent = { type: 'image_url', image_url: { url: imageUrl, detail } };
    } else {
      throw new Error('Either imagePath or imageUrl required');
    }

    try {
      const response = await this.client.chat.completions.create({
        model,
        messages: [{ role: 'user', content: [{ type: 'text', text: prompt }, imageContent] }],
        max_tokens: this.generationConfig.max_tokens
      });

      return { text: response.choices[0].message.content, usage: response.usage };
    } catch (error) {
      return this.handleError(error);
    }
  }

  /** Tool / function calling */
  async generateWithTools(options) {
    await this.rateLimit();
    const { prompt, tools, model, systemPrompt } = options;

    const messages = [];
    if (systemPrompt) messages.push({ role: 'system', content: systemPrompt });
    messages.push({ role: 'user', content: prompt });

    const openaiTools = tools.map(t => ({
      type: 'function',
      function: { name: t.name, description: t.description, parameters: t.parameters }
    }));

    try {
      const response = await this.client.chat.completions.create({
        model: model || this.defaultModel,
        messages,
        tools: openaiTools,
        tool_choice: 'auto'
      });

      const msg = response.choices[0].message;
      const toolCall = msg.tool_calls?.[0];

      return {
        text: msg.content || '',
        toolCall: toolCall ? {
          name: toolCall.function.name,
          args: JSON.parse(toolCall.function.arguments)
        } : null,
        usage: response.usage
      };
    } catch (error) {
      return this.handleError(error);
    }
  }

  /** Image generation with Aurora */
  async generateImage(options) {
    await this.rateLimit();
    const { prompt, model = 'aurora', n = 1, responseFormat = 'url' } = options;

    try {
      const response = await this.client.images.generate({
        model, prompt, n, response_format: responseFormat
      });
      return response.data.map(img => ({ url: img.url, b64: img.b64_json }));
    } catch (error) {
      return this.handleError(error);
    }
  }

  handleError(error) {
    const response = { error: true, message: error.message, type: error.constructor.name };

    if (error.status === 401) {
      response.type = 'AuthenticationError';
      response.message = 'Invalid API key. Check XAI_API_KEY';
    } else if (error.status === 429) {
      response.type = 'RateLimitError';
      response.message = 'Rate limit exceeded. Please wait and retry';
    }

    if (this.settings.logging.logRequests) console.error('Grok API Error:', response);
    throw response;
  }

  getAvailableModels() { return this.modelConfigs.models; }
  getRecommendedModel(t) { return this.modelConfigs.modelSelection[t] || this.defaultModel; }

  async listModels() {
    try {
      const response = await this.client.models.list();
      return response.data;
    } catch (error) {
      return this.handleError(error);
    }
  }
}

module.exports = { GrokClient };
